Migrate to python 3 + Invoke (#61)
jolicode/docker-starter: UPGRADE-3.0.md

How to upgrade from 2.x to 3.0
==============================

> This guide will cover the migration from fabric to invoke.

Reviewer: ```suggestion
This guide will cover the migration from Fabric to Invoke.
```

At anytime, you can refer the final form of the starter kit by looking
at the [main repository](https://github.com/jolicode/docker-starter)

> 1. Rename the `fabfile.py` to `task.py`:

Reviewer 2: ```suggestion
1. Rename the `fabfile.py` to `tasks.py`:
```

    ```bash
    mv fabfile.py tasks.py
    ```

1. Extract the configuration to a dedicated file `invoke.py`:

>     1. The variable are not longer prefixed with `env.`:

Reviewer 2: ```suggestion
    1. The variables are not longer prefixed with `env.`:
```

        * `env.project_name` => `project_name`;
        * `env.root_domain` => `root_domain`;
        * `env.project_directory` => `project_directory`;
        * `env.extra_domains` => `extra_domains`.

>     1. Append the the following content to the file:

Reviewer 2: ```suggestion
    1. Append the following content to the file:
```

                shellProjectName = run('echo %PROJECT_NAME%', hide=True).stdout
                if (shellProjectName != config['project_name']):
                    domains = '`' + '`, `'.join([config['root_domain']] + config['extra_domains']) + '`'
                    print(Fore.RED + 'Env vars not set (Windows detected)')
                    print(Fore.YELLOW + 'You must manually set environment variables on Windows:')
                    print('$Env:PROJECT_NAME="%s"' % config['project_name'])
                    print('$Env:PROJECT_DIRECTORY="%s"' % config['project_directory'])
                    print('$Env:PROJECT_ROOT_DOMAIN="%s"' % config['root_domain'])
                    print('$Env:PROJECT_DOMAINS="%s"' % domains)
                    sys.exit(0)

            if not config['power_shell']:
                config['user_id'] = int(run('id -u', hide=True).stdout)

            if config['user_id'] > 256000:
                config['user_id'] = 1000

            return config

        locals().update(__extract_runtime_configuration(locals()))
        ```

    1. You can remove the similar part from `tasks.py` because `invoke.py` will be automatically imported.

>         Don't forget to `remove set_local_configuration()` function and its function call.

Author: ```suggestion
        Don't forget to remove `set_local_configuration()` function and its function call.
```

            else:
                docker_compose_run(c, command, no_deps=no_deps)

        def docker_compose_run(c, command_name, service="builder", user="app", no_deps=False, workdir=None, port_mapping=False):
            args = [
                'run',
                '--rm',
                '-u %s' % quote(user),
            ]

            if no_deps:
                args.append('--no-deps')

            if port_mapping:
                args.append('--service-ports')

            if workdir is not None:
                args.append('-w %s' % quote(workdir))

            docker_compose(c, '%s %s /bin/sh -c "exec %s"' % (
                ' '.join(args),
                quote(service),
                command_name
            ))

        def docker_compose(c, command_name):
            domains = '`' + '`, `'.join([c.root_domain] + c.extra_domains) + '`'

            env = {
                'PROJECT_NAME': c.project_name,
                'PROJECT_DIRECTORY': c.project_directory,
                'PROJECT_ROOT_DOMAIN': c.root_domain,
                'PROJECT_DOMAINS': domains,
            }

            cmd = 'docker-compose -p %s %s %s' % (
                c.project_name,

>                 ' '.join('-f \'' + c.root_dir + '/infrastructure/docker/' + file + '\'' for file in c.docker_compose_files),

Reviewer 3: Something doesn't work well here, I get this error:

> .OSError: [Errno 22] Invalid argument: ".\\'C:\\Users\\mobma\\Dev\\docker-starter/infrastructure/docker/docker-compose.builder.yml'"

The root_dir is quite full. But if I replace it with ./ I still get an escaped `'`:

> .FileNotFoundError: [Errno 2] No such file or directory: ".\\'./infrastructure/docker/docker-compose.builder.yml'"

Using directly this works:

```python
' '.join('-f infrastructure/docker/' + file + '' for file in c.docker_compose_files),
```

Author: Could you try to run the command with `-e` ?
```
inv -e up
```

Reviewer 3: Here is the result:

```
docker-compose -p app -f 'C:\Users\mobma\Dev\docker-starter/infrastructure/docker/docker-compose.builder.yml' -f 'C:\Users\mobma\Dev\docker-starter/infrastructure/docker/docker-compose.yml' -f 'C:\Users\mobma\Dev\docker-starter/infrastructure/docker/docker-compose.worker.yml' build --build-arg PROJECT_NAME=app --build-arg USER_ID=1000 php-base
```

Doesn't look so bad.

Reviewer 3: What is strange is that I can run this command directly from Powershell and it works.

Author: For the record: 

The running command displayed by invoke is good :+1: 

```
docker-compose -p app -f 'C:\Users\greg\dev\docker-starter/infrastructure/docker/docker-compose.yml' -f 'C:\Users\greg\dev\docker-starter/infrastructure/docker/docker-compose.worker.yml' ps
```

If we run it manually it works.

But when it goes through invoke it does not work anymore and we got the error:

> .OSError: [Errno 22] Invalid argument: ".\'C:\Users\greg\dev\docker-starter/infrastructure/docker/docker-compose.builder.yml'"

I tried to follow the code path

1. invoke: https://github.com/pyinvoke/invoke/blob/master/invoke/runners.py#L1271 
2. Python: https://docs.python.org/3/library/subprocess.html#subprocess.Popen

So there is a mismatch here. Popen expects an array, but invoke give a string.

Anyway, regular escaping seems to work, but not in the case of a PATH. I suspect a wizard that "escape" path on windows. 

We must surround PATH with quote to avoid any issues.
So I change from single quote `'` to double quote `"` and now it works on windows and linux

If someone with strong python knowledge could explain to me what is happening here, I will be very happy.

Reviewer 4: Single and double quotes are not interchangeable in powershell or batch. Batch requires double quotes with quite complex escaping rules (e.g. https://ss64.com/nt/syntax-esc.html). I'm not too familiar with powershell, but I believe they use back-ticks for quoting arguments.

            )

            c.run(cmd, pty=True, env=env)
        ```

>     1. Use the context in all task

Reviewer 2: ```suggestion
    1. Use the context in all tasks
```

        Invoke needs a context (named `c`) in every single task. This context:

        * is configured for each task;
        * contains the configuration;
        * contains some methods to execute commands.

>         You will need to update all task signatures, and all task function calls to pass the context:

Author: ```suggestion
        You will need to update all tasks signature, and all tasks function call to pass the context:
```

            machine_running = local('dinghy status', capture=True)
        ```

        After:

        ```py
        @task
        def start(c):
            machine_running = c.local('dinghy status').stdout
        ```

    1. Add `Builder` class and use it:

        The Builder class is the new `@with_builder` decorator

        ```py
        class Builder:
            def __init__(self, c):
                self.c = c

            def __enter__(self):
                self.docker_compose_files = self.c.docker_compose_files
                self.c.docker_compose_files = ['docker-compose.builder.yml'] + self.docker_compose_files

            def __exit__(self, type, value, traceback):
                self.c.docker_compose_files = self.docker_compose_files
        ```

        Before:

        ```py
        @task
        @with_builder
        def install():
            docker_compose_run('composer install -n --prefer-dist --optimize-autoloader')
        ```

        After:

        ```py
        @task
        def install(c):
            with Builder(c):
                docker_compose_run(c, 'composer install -n --prefer-dist --optimize-autoloader')
        ```

    1. Update colors

>         Fabric povided nice helper for colors. Now we use colorama

Reviewer: ```suggestion
        Fabric provided nice helper for colors. Now we use colorama
```

> 

Author: ```suggestion

```
